cloud/pkg/devicecontroller/controller/upstream.go

```diff
@@ -96,15 +102,79 @@ func (uc *UpstreamController) dispatchMessage() {
 		klog.Infof("Message: %s, resource type is: %s", msg.GetID(), resourceType)
 
 		switch resourceType {
-		case constants.ResourceTypeTwinEdgeUpdated:
+		case messagelayer.ResourceTypeTwinEdgeUpdated:
 			uc.deviceStatusChan <- msg
-		case constants.ResourceTypeMembershipDetail:
+		case messagelayer.ResourceTypeStateEdgeUpdated:
+			uc.deviceStateChan <- msg
+		case messagelayer.ResourceTypeMembershipDetail:
 		default:
 			klog.Warningf("Message: %s, with resource type: %s not intended for device controller", msg.GetID(), resourceType)
 		}
 	}
 }
 
+func (uc *UpstreamController) updateDeviceState() {
+	for {
+		select {
+		case <-beehiveContext.Done():
+			klog.Info("Stop updateDeviceState")
+			return
+		case msg := <-uc.deviceStateChan:
+			klog.Infof("Message: %s, operation is: %s, and resource is: %s", msg.GetID(), msg.GetOperation(), msg.GetResource())
+			deviceStateEdge, err := uc.unmarshalDeviceStateMessage(msg)
+			if err != nil {
+				klog.Warningf("Unmarshall failed due to error %v", err)
+				continue
+			}
+			deviceID, err := messagelayer.GetDeviceID(msg.GetResource())
+			if err != nil {
+				klog.Warning("Failed to get device id")
+				continue
+			}
+			device, ok := uc.dc.deviceManager.Device.Load(deviceID)
+			if !ok {
+				klog.Warningf("Device %s does not exist in downstream controller", deviceID)
+				continue
+			}
+			cacheDevice, ok := device.(*v1alpha2.Device)
+			if !ok {
+				klog.Warning("Failed to assert to CacheDevice type")
+				continue
+			}
+			deviceState := &DeviceStatus{Status: cacheDevice.Status}
+
+			deviceState.Status.State = deviceStateEdge.Device.State
+			cacheDevice.Status.State = deviceStateEdge.Device.State
+
+			state := strings.ToLower(cacheDevice.Status.State)
+			// if the device is online, update the last online time to k8s
+			if state == "ok" || state == "online" {
+				deviceState.Status.LastOnline = deviceStateEdge.Device.LastOnline
+				cacheDevice.Status.LastOnline = deviceStateEdge.Device.LastOnline
+			}
+
+			body, err := json.Marshal(deviceState)
+			if err != nil {
+				klog.Errorf("Failed to marshal device state %v", deviceState)
+				continue
+			}
+			err = uc.crdClient.DevicesV1alpha2().RESTClient().Patch(MergePatchType).Namespace(cacheDevice.Namespace).Resource(ResourceTypeDevices).Name(deviceID).Body(body).Do(context.Background()).Error()
```

Comment:
please use the generated client to update the state

Reply:
done